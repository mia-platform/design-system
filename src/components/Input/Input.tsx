/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  ChangeEvent,
  ReactElement,
  useCallback,
  useMemo,
  useState,
} from 'react'
import { Input as AntInput } from 'antd'

import { AddonType, Type } from './types'
import { BaseInput, defaults as baseInputDefaults } from '../BaseInput/BaseInput'
import { InputAddon, InputAddonProps } from './InputAddon.tsx'
import { Icon } from '../Icon'
import { InputProps } from './props'
import { isObject } from '../../utils/object.ts'
import styles from './input.module.css'

export const defaults = {
  ...baseInputDefaults,
  htmlType: Type.Text,
  valuePropName: 'value',
}

const DEFAULT_ICON_SIZE = 12 as never

type AddonPosition = 'before' | 'after'

const getAddonDefaultValue = (position: AddonPosition, props?: InputAddonProps) : Record<string, unknown> => {
  return props && (props.value !== undefined || props.defaultValue !== undefined)
    ? { [props.name || position]: props.value || props.defaultValue }
    : {}
}

const useInputValue = (
  value: Record<string, unknown> | string | undefined,
  valuePropName: string,
  { before, after }: Record<AddonPosition, InputAddonProps | undefined>
):
  [
    Record<string, unknown>,
    React.Dispatch<React.SetStateAction<Record<string, unknown>>>
  ] => {
  return useState({
    ...(isObject(value) ? value : { [valuePropName]: value || '' }),
    ...getAddonDefaultValue('before', before),
    ...getAddonDefaultValue('after', after),
  })
}

/**
 * A UI element to insert text content in a form.
 *
 * @link https://ant.design/components/input
 * @returns {Input} Input component
 */
export const Input = (
  {
    id,
    appearance = defaults.type,
    type = defaults.htmlType,
    value,
    defaultValue,
    valuePropName = defaults.valuePropName,
    isDisabled,
    isReadOnly,
    isFullWidth = defaults.isFullWidth,
    isError,
    inputRef,
    onChange,
    placeholder,
    iconLeft,
    autoFocus,
    iconRight,
    allowClear,
    maxLength,
    minLength,
    addonAfter: addonAfterProp,
    addonBefore: addonBeforeProp,
  }: InputProps
) : ReactElement => {
  const [val, setVal] = useInputValue(value || defaultValue, valuePropName, {
    before: addonBeforeProp,
    after: addonAfterProp,
  })

  const handleChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    const { value: nextValue } = event.target
    const nextVal = { ...val, [valuePropName]: nextValue }
    setVal(nextVal)
    if (onChange) {
      onChange(event, addonBeforeProp || addonAfterProp ? nextVal : nextValue)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [addonAfterProp, addonBeforeProp, onChange, valuePropName])

  const renderAddon = useCallback((
    position: AddonPosition,
    {
      value: addonValue,
      onChange: onChangeAddon,
      disabled: addonDisabled,
      name: addonName,
      ...rest
    }: InputAddonProps
  ) => {
    return (
      <InputAddon
        {...rest}
        isDisabled={addonDisabled === undefined ? isDisabled : addonDisabled}
        value={addonValue !== undefined ? addonValue : val[addonName || position]}
        onChange={(nextValue) => {
          const nextVal = { ...val, [addonName || position]: nextValue }
          setVal(nextVal)
          if (onChange) {
            onChange(undefined, nextVal)
          }
          if (onChangeAddon) {
            onChangeAddon(nextValue)
          }
        }}
      />
    )
  }, [isDisabled, onChange, setVal, val])

  const addonBefore = useMemo(() => (
    addonBeforeProp && renderAddon('before', addonBeforeProp)
  ), [addonBeforeProp, renderAddon])

  const addonAfter = useMemo(() => (
    addonAfterProp && renderAddon('after', addonAfterProp)
  ), [addonAfterProp, renderAddon])

  const getValue = useCallback((strOrObj?: string | Record<string, unknown>) => {
    return isObject(strOrObj) ? String(strOrObj[valuePropName]) : strOrObj
  }, [valuePropName])

  const inputValue = useMemo(() => {
    return getValue(value)
  }, [getValue, value])

  const inputDefaultValue = useMemo(() => {
    return getValue(defaultValue)
  }, [getValue, defaultValue])

  return (
    <BaseInput
      addonAfter={addonAfter}
      addonBefore={addonBefore}
      allowClear={allowClear}
      appearance={appearance}
      autoFocus={autoFocus}
      className={styles.input}
      component={AntInput}
      defaultValue={inputDefaultValue}
      id={id}
      inputRef={inputRef}
      isDisabled={isDisabled}
      isError={isError}
      isFullWidth={isFullWidth}
      isReadOnly={isReadOnly}
      maxLength={maxLength}
      minLength={minLength}
      placeholder={placeholder}
      prefix={iconLeft && <Icon component={iconLeft} size={DEFAULT_ICON_SIZE} />}
      suffix={iconRight && <Icon component={iconRight} size={DEFAULT_ICON_SIZE} />}
      type={type}
      value={inputValue}
      onChange={handleChange}
    />
  )
}

Input.Appearance = BaseInput.Appearance
Input.Type = Type
Input.AddonType = AddonType
