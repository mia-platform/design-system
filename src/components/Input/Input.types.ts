
export enum Type {
  Outlined = 'outlined',
  Borderless = 'borderless'
}

export enum HTMLType {
  Text = 'text',
  Textarea = 'textarea',
  Number = 'number',
  Password = 'password',
  Search = 'search',
  Hidden = 'hidden',
}
