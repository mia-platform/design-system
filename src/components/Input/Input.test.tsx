/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { FiSearch } from 'react-icons/fi'

import { fireEvent, render, screen } from '../../test-utils'
import { Input } from './Input'

const textAddonValue = 'test'

const checkboxAddonLabel = 'test'

const selectAddonOptions = [
  ...Array(5).keys(),
].map((id) => ({
  value: `value ${id + 1}`,
}))

describe('Input Component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('renders correctly', () => {
    const { asFragment } = render(<Input />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('With left icon renders correctly', () => {
    const { asFragment } = render(<Input iconLeft={FiSearch} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('With right icon renders correctly', () => {
    const { asFragment } = render(<Input iconRight={FiSearch} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AllowClear renders correctly', () => {
    const { asFragment } = render(<Input allowClear defaultValue="text" />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('TextArea renders correctly', () => {
    const { asFragment } = render(<Input type={Input.Type.Textarea} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Search renders correctly', () => {
    const { asFragment } = render(<Input type={Input.Type.Search} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Password renders correctly', () => {
    const { asFragment } = render(<Input type={Input.Type.Number} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Hidden renders correctly', () => {
    const { asFragment } = render(<Input type={Input.Type.Hidden} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonBeforeText renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'text', value: textAddonValue }} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonBeforeSelect renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'select', options: selectAddonOptions }} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonBeforeCheckbox renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'checkbox', label: checkboxAddonLabel }} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonAfterText renders correctly', () => {
    const { asFragment } = render(<Input addonAfter={{ type: 'text', value: textAddonValue }} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonAfterSelect renders correctly', () => {
    const { asFragment } = render(<Input addonAfter={{ type: 'select', options: selectAddonOptions }} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonAfterCheckbox renders correctly', () => {
    const { asFragment } = render(<Input addonAfter={{ type: 'checkbox', label: checkboxAddonLabel }} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonTextDisabled renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'text', value: textAddonValue }} isDisabled={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('AddonSelectDisabled renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'select', options: selectAddonOptions }} isDisabled={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  // Test per la storia AddonCheckboxDisabled
  test('AddonCheckboxDisabled renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'checkbox', label: checkboxAddonLabel }} isDisabled={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  // Test per la storia AddonCheckboxDisabledChecked
  test('AddonCheckboxDisabledChecked renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ defaultValue: true, type: 'checkbox', label: checkboxAddonLabel }} isDisabled={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  // Test per la storia AddonTextError
  test('AddonTextError renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'text', value: textAddonValue }} isError={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  // Test per la storia AddonSelectError
  test('AddonSelectError renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'select', options: selectAddonOptions }} isError={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  // Test per la storia AddonCheckboxError
  test('AddonCheckboxError renders correctly', () => {
    const { asFragment } = render(<Input addonBefore={{ type: 'checkbox', label: checkboxAddonLabel }} isError={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Should call event handler on Change', async() => {
    const onChange = jest.fn()
    const text = 'text'

    render(<Input onChange={onChange} />)

    const input = screen.getByRole<HTMLInputElement>('textbox')

    fireEvent.change(input, { target: { value: text } })

    expect(onChange).toHaveBeenCalled()
    expect(input.value).toBe(text)
  })

  test('Allow clear should clear input', () => {
    const text = 'text'

    render(<Input allowClear defaultValue={text} />)

    const input = screen.getByRole<HTMLInputElement>('textbox')
    const clearIcon = screen.getByRole('img', { name: 'close-circle' })

    fireEvent.click(clearIcon)

    expect(input.value).toBe('')
  })
})
