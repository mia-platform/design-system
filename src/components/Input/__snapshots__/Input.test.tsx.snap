// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component AddonAfterCheckbox renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <input
        class="mia-platform-input mia-platform-input-outlined"
        type="text"
        value=""
      />
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonCheckbox"
        >
          <div
            class="checkboxComponent"
          >
            <div
              class="checkboxTextWrapper"
            >
              <label
                class="mia-platform-checkbox-wrapper"
              >
                <span
                  class="mia-platform-checkbox ant-wave-target"
                >
                  <input
                    class="mia-platform-checkbox-input"
                    type="checkbox"
                  />
                  <span
                    class="mia-platform-checkbox-inner"
                  />
                </span>
              </label>
              test
            </div>
          </div>
        </div>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonAfterSelect renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <input
        class="mia-platform-input mia-platform-input-outlined"
        type="text"
        value=""
      />
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonSelect"
        >
          <div
            class="mia-platform-select mia-platform-select-outlined input select mia-platform-select-single mia-platform-select-show-arrow"
          >
            <div
              class="mia-platform-select-selector"
            >
              <span
                class="mia-platform-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="mia-platform-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="mia-platform-select-selection-placeholder"
              />
            </div>
            <span
              aria-hidden="true"
              class="mia-platform-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <svg
                color="currentColor"
                fill="currentColor"
                height="16"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                style="color: currentColor;"
                viewBox="0 0 256 256"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
                />
              </svg>
            </span>
          </div>
        </div>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonAfterText renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <input
        class="mia-platform-input mia-platform-input-outlined"
        type="text"
        value=""
      />
      <span
        class="mia-platform-input-group-addon"
      >
        test
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonBeforeCheckbox renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonCheckbox"
        >
          <div
            class="checkboxComponent"
          >
            <div
              class="checkboxTextWrapper"
            >
              <label
                class="mia-platform-checkbox-wrapper"
              >
                <span
                  class="mia-platform-checkbox ant-wave-target"
                >
                  <input
                    class="mia-platform-checkbox-input"
                    type="checkbox"
                  />
                  <span
                    class="mia-platform-checkbox-inner"
                  />
                </span>
              </label>
              test
            </div>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-outlined"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonBeforeSelect renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonSelect"
        >
          <div
            class="mia-platform-select mia-platform-select-outlined input select mia-platform-select-single mia-platform-select-show-arrow"
          >
            <div
              class="mia-platform-select-selector"
            >
              <span
                class="mia-platform-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="mia-platform-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="mia-platform-select-selection-placeholder"
              />
            </div>
            <span
              aria-hidden="true"
              class="mia-platform-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <svg
                color="currentColor"
                fill="currentColor"
                height="16"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                style="color: currentColor;"
                viewBox="0 0 256 256"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
                />
              </svg>
            </span>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-outlined"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonBeforeText renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        test
      </span>
      <input
        class="mia-platform-input mia-platform-input-outlined"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonCheckboxDisabled renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-disabled mia-platform-input-group-wrapper-outlined input fullWidth disabled input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonCheckbox"
        >
          <div
            class="checkboxComponent disabled"
          >
            <div
              class="checkboxTextWrapper"
            >
              <label
                class="mia-platform-checkbox-wrapper mia-platform-checkbox-wrapper-disabled"
              >
                <span
                  class="mia-platform-checkbox ant-wave-target mia-platform-checkbox-disabled"
                >
                  <input
                    class="mia-platform-checkbox-input"
                    disabled=""
                    type="checkbox"
                  />
                  <span
                    class="mia-platform-checkbox-inner"
                  />
                </span>
              </label>
              test
            </div>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-disabled mia-platform-input-outlined"
        disabled=""
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonCheckboxDisabledChecked renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-disabled mia-platform-input-group-wrapper-outlined input fullWidth disabled input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonCheckbox"
        >
          <div
            class="checkboxComponent disabled"
          >
            <div
              class="checkboxTextWrapper"
            >
              <label
                class="mia-platform-checkbox-wrapper mia-platform-checkbox-wrapper-disabled"
              >
                <span
                  class="mia-platform-checkbox ant-wave-target mia-platform-checkbox-checked mia-platform-checkbox-disabled"
                >
                  <input
                    checked=""
                    class="mia-platform-checkbox-input"
                    disabled=""
                    type="checkbox"
                  />
                  <span
                    class="mia-platform-checkbox-inner"
                  />
                </span>
              </label>
              test
            </div>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-disabled mia-platform-input-outlined"
        disabled=""
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonCheckboxError renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined mia-platform-input-group-wrapper-status-error input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonCheckbox"
        >
          <div
            class="checkboxComponent"
          >
            <div
              class="checkboxTextWrapper"
            >
              <label
                class="mia-platform-checkbox-wrapper"
              >
                <span
                  class="mia-platform-checkbox ant-wave-target"
                >
                  <input
                    class="mia-platform-checkbox-input"
                    type="checkbox"
                  />
                  <span
                    class="mia-platform-checkbox-inner"
                  />
                </span>
              </label>
              test
            </div>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-outlined mia-platform-input-status-error"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonSelectDisabled renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-disabled mia-platform-input-group-wrapper-outlined input fullWidth disabled input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonSelect"
        >
          <div
            class="mia-platform-select mia-platform-select-outlined input disabled select disabled mia-platform-select-single mia-platform-select-show-arrow mia-platform-select-disabled"
          >
            <div
              class="mia-platform-select-selector"
            >
              <span
                class="mia-platform-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="mia-platform-select-selection-search-input"
                  disabled=""
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="mia-platform-select-selection-placeholder"
              />
            </div>
            <span
              aria-hidden="true"
              class="mia-platform-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <svg
                color="currentColor"
                fill="currentColor"
                height="16"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                style="color: currentColor;"
                viewBox="0 0 256 256"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
                />
              </svg>
            </span>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-disabled mia-platform-input-outlined"
        disabled=""
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonSelectError renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined mia-platform-input-group-wrapper-status-error input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        <div
          class="inputAddonSelect"
        >
          <div
            class="mia-platform-select mia-platform-select-outlined input select mia-platform-select-single mia-platform-select-show-arrow"
          >
            <div
              class="mia-platform-select-selector"
            >
              <span
                class="mia-platform-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="mia-platform-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="mia-platform-select-selection-placeholder"
              />
            </div>
            <span
              aria-hidden="true"
              class="mia-platform-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <svg
                color="currentColor"
                fill="currentColor"
                height="16"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                style="color: currentColor;"
                viewBox="0 0 256 256"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
                />
              </svg>
            </span>
          </div>
        </div>
      </span>
      <input
        class="mia-platform-input mia-platform-input-outlined mia-platform-input-status-error"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonTextDisabled renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-disabled mia-platform-input-group-wrapper-outlined input fullWidth disabled input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        test
      </span>
      <input
        class="mia-platform-input mia-platform-input-disabled mia-platform-input-outlined"
        disabled=""
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AddonTextError renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-group-wrapper mia-platform-input-group-wrapper-outlined mia-platform-input-group-wrapper-status-error input fullWidth input"
  >
    <span
      class="mia-platform-input-wrapper mia-platform-input-group"
    >
      <span
        class="mia-platform-input-group-addon"
      >
        test
      </span>
      <input
        class="mia-platform-input mia-platform-input-outlined mia-platform-input-status-error"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component AllowClear renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-affix-wrapper mia-platform-input-outlined input fullWidth input"
  >
    <input
      class="mia-platform-input"
      type="text"
      value="text"
    />
    <span
      class="mia-platform-input-suffix"
    >
      <span
        class="mia-platform-input-clear-icon"
        role="button"
        tabindex="-1"
      >
        <span
          aria-label="close-circle"
          class="anticon anticon-close-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close-circle"
            fill="currentColor"
            fill-rule="evenodd"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
            />
          </svg>
        </span>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component Hidden renders correctly 1`] = `
<DocumentFragment>
  <input
    class="mia-platform-input mia-platform-input-outlined input fullWidth input"
    type="hidden"
    value=""
  />
</DocumentFragment>
`;

exports[`Input Component Password renders correctly 1`] = `
<DocumentFragment>
  <input
    class="mia-platform-input mia-platform-input-outlined input fullWidth input"
    type="number"
    value=""
  />
</DocumentFragment>
`;

exports[`Input Component Search renders correctly 1`] = `
<DocumentFragment>
  <input
    class="mia-platform-input mia-platform-input-outlined input fullWidth input"
    type="search"
    value=""
  />
</DocumentFragment>
`;

exports[`Input Component TextArea renders correctly 1`] = `
<DocumentFragment>
  <input
    class="mia-platform-input mia-platform-input-outlined input fullWidth input"
    type="textarea"
    value=""
  />
</DocumentFragment>
`;

exports[`Input Component With left icon renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-affix-wrapper mia-platform-input-outlined input fullWidth input"
  >
    <span
      class="mia-platform-input-prefix"
    >
      <svg
        color="currentColor"
        fill="none"
        height="12"
        role="img"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: currentColor;"
        viewBox="0 0 24 24"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <line
          x1="21"
          x2="16.65"
          y1="21"
          y2="16.65"
        />
      </svg>
    </span>
    <input
      class="mia-platform-input"
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;

exports[`Input Component With right icon renders correctly 1`] = `
<DocumentFragment>
  <span
    class="mia-platform-input-affix-wrapper mia-platform-input-outlined input fullWidth input"
  >
    <input
      class="mia-platform-input"
      type="text"
      value=""
    />
    <span
      class="mia-platform-input-suffix"
    >
      <svg
        color="currentColor"
        fill="none"
        height="12"
        role="img"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: currentColor;"
        viewBox="0 0 24 24"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <line
          x1="21"
          x2="16.65"
          y1="21"
          y2="16.65"
        />
      </svg>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input Component renders correctly 1`] = `
<DocumentFragment>
  <input
    class="mia-platform-input mia-platform-input-outlined input fullWidth input"
    type="text"
    value=""
  />
</DocumentFragment>
`;
